util.c

```diff
@@ -1145,3 +1149,116 @@ void drop_cookie_secret(struct nsd* nsd)
 	              , NSD_COOKIE_SECRET_SIZE);
 	nsd->cookie_count -= 1;
 }
+
+static
+int cookie_secret_file_read(nsd_type* nsd) {
+	cookie_secret_type cookie_secrets[NSD_COOKIE_HISTORY_SIZE];
+	char secret[NSD_COOKIE_SECRET_SIZE * 2 + 2/*'\n' and '\0'*/];
+	FILE* f;
+	size_t count = 0;
+	const char* fn;
+
+	if(!(fn = cookie_secret_file(nsd->options)))
+		return 1; /* Explicitely disabled with empty filename */
+
+	else if((f = fopen(fn, "r")) != NULL)
+		; /* pass */
+
+	/* a non-existing cookie file is not an error */
+	else if(errno != ENOENT) {
+		log_msg( LOG_ERR
+		       , "error reading cookie secret file \"%s\": \"%s\""
+		       , fn, strerror(errno));
+		return 0;
+	}
+	else if(nsd->options->cookie_secret_file != NULL /* explicit name */
+	     ||!(f = fopen((fn = CONFIGDIR"/nsd_cookiesecrets.txt"), "r")))
```

Comment:
Falling back to `nsd_cookiesecrets.txt` in this function doesn't seem right. I think it's preferred to handle that in the configuration parser (or shortly after). Maybe do something like if `cookie_secret_file` is still `NULL` after parsing, set it to the previous default? That way behavior is more deterministic. Also would avoid the `cookie_secret_file` function, which seems to return `COOKIESECRETSFILE` if the `cookie_secret_file` is `NULL`. So, basically `fn` is only `NULL` if the `cookie_secret_file` is not `NULL` but empty, in which case `NULL` is returned. This seems a very convoluted way of avoiding checking whether the member is `NULL` just after parsing and setting right then(?)

Reply:
Okay, addressed in [97fec31](https://github.com/NLnetLabs/nsd/pull/378/commits/97fec31c7d2ea184b1123bea6139f325ecf9de9f)